src/transport/ExchangeMgr.h

```diff
@@ -0,0 +1,276 @@
+        /**< Deprecated alias for \c kSendFlag_DefaultMulticastSourceAddress */
+    };
+
+    bool IsInitiator() const;
+    bool IsResponseExpected() const;
+    void SetInitiator(bool inInitiator);
+    void SetResponseExpected(bool inResponseExpected);
+
+    CHIP_ERROR SendMessage(uint32_t protocolId, uint8_t msgType, System::PacketBuffer * msgPayload, uint16_t sendFlags = 0,
+                           void * msgCtxt = nullptr);
+    CHIP_ERROR SendCommonNullMessage();
+
+    void TeardownTrickleRetransmit();
+
+    /**
+     * This function is the protocol callback for handling a received CHIP message.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     *  @param[in]    packetHeader  A reference to the PacketHeader object.
+     *
+     *  @param[in]    protocolId    The protocol identifier of the received message.
+     *
+     *  @param[in]    msgType       The message type of the corresponding profile.
+     *
+     *  @param[in]    payload       A pointer to the PacketBuffer object holding the message payload.
+     */
+    typedef void (*MessageReceiveFunct)(ExchangeContext * ec, const PacketHeader & packetHeader, uint32_t protocolId,
+                                        uint8_t msgType, System::PacketBuffer * payload);
+    MessageReceiveFunct OnMessageReceived;
+
+    /**
+     * This function is the protocol callback to invoke when the timeout for the receipt
+     * of a response message has expired.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     */
+    typedef void (*ResponseTimeoutFunct)(ExchangeContext * ec);
+    ResponseTimeoutFunct OnResponseTimeout;
+
+    /**
+     * This function is the protocol callback to invoke when the timeout for the retransmission
+     * of a previously sent message has expired.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     */
+    typedef void (*RetransmissionTimeoutFunct)(ExchangeContext * ec);
+    RetransmissionTimeoutFunct OnRetransmissionTimeout;
+
+    /**
+     * Type of key error message handling function.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     *  @param[in]    keyErr        The CHIP_ERROR type that was reported in the key error message.
+     *
+     */
+    typedef void (*KeyErrorFunct)(ExchangeContext * ec, CHIP_ERROR keyErr);
+
+    /**
+     * This function is the protocol callback to invoke when key error message has been received
+     * from the peer.
+     */
+    KeyErrorFunct OnKeyError;
+
+    void CancelRetrans();
+
+    /*
+     * in order to use reference counting (see refCount below)
+     * we use a hold/free paradigm where users of the exchange
+     * can hold onto it while it's out of their direct control
+     * to make sure it isn't closed before everyone's ready.
+     */
+    void AddRef();
```

Comment:
Let's use `ReferenceCounted` template. It's well tested, and already being used by other parts of `CHIP`.

Reply:
I can not use the ReferenceCounted since there are some extra stuff to do within Retain and Release. For example, in case the reference counter decreased to zero, the ExchangeContext can not free itself immediately. It will handle over to ExchangeManager and tell it 'I am done'. ExchangeManager will handle the remaining cleanup.

This might happen for example, in the last message in an exchange. The protocol received a message, and is sending back the last message in that exchange. From its perspective, it's 'done' with that exchange, but the exchange may still be active for a short while after while the ExchangeManager attempts to deliver that reliably. The protocol 'transfers' the ExchangeContext object to the ExchangeManager, which then handles the final clean-up duties based on getting it reliably sent, before finally cleaning up that object.

Reply:
This explanation would be a great addition in the comments on those fields in the class.